chore(tracing): encapsulate agent-specific writer functionality

emmettbutler opened the pull request:
This change facilitates the creation of new `*Writer` classes by encapsulating the behavior of the `AgentWriter` that is specific to communication with the Datadog Agent in its own class. This leaves most of the writer's functionality, which could apply to any trace writing workload that uses HTTP, in the new `HTTPWriter` class. This change builds on the approach from https://github.com/DataDog/dd-trace-py/pull/3690.

https://github.com/DataDog/dd-trace-py/pull/5471 builds on this change to implement a new `CIVisibilityWriter` class.

This change only affects code organization, not functionality, so existing tests are sufficient to cover the risk.

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/contributing.html#Release-Note-Guidelines) are followed.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] PR description includes explicit acknowledgement/acceptance of the performance implications of this PR as reported in the benchmarks PR comment.

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.

mabdinur reviewed (commented):
An alternative approach is to create an `_HTTPClient` class that encapsulates all logic required to encode and send an http requests containing traces. We can use this client in both the `AgentWriter` and the proposed `CIVisiblityWriter`.  We already use this pattern in the [remote config worker](https://github.com/DataDog/dd-trace-py/blob/f8654e00036bb58bbf5c562ad0178239a043799d/ddtrace/internal/remoteconfig/worker.py#L45) and the [telemetry writer](https://github.com/DataDog/dd-trace-py/blob/v1.11.1/ddtrace/internal/telemetry/writer.py#L132). 

But this pattern is not used in [dynamic instrumentation](https://github.com/DataDog/dd-trace-py/blob/v1.11.1/ddtrace/debugging/_debugger.py#L162), or the [profiler](https://github.com/DataDog/dd-trace-py/blob/f1069ccb0a2a9ae4c78dbfe1a6847da520b8440b/ddtrace/profiling/profiler.py#L181) so I am fine with the current approach.

emmettbutler commented:
@mabdinur good idea, I'll probably do that in a subsequent pull request. I've addressed all of your suggestions and this change is ready for another review.

Yun-Kim commented on ddtrace/internal/writer.py:
```diff
@@ -625,11 +494,215 @@ def _stop_service(
+        )
+
+        endpoint = "%s/traces" % self._api_version
+        headers = {
```
@emmettbutler from what I can tell, it looks like we never use the provided `headers` argument and instead always override them to be this, while previously it would update the provided `headers` argument. Is this intentional?

emmettbutler replied:
Good catch, that was unintentional. I'll fix it and add a regression test.

emmettbutler replied:
https://github.com/DataDog/dd-trace-py/pull/5483

Yun-Kim reviewed (commented):
Just one small question, but 💯
